kweet/test/KweetApplicationTestLegacy.kt

```diff
@@ -0,0 +1,122 @@
+        every { dao.getKweet(1) } returns Kweet(1, "user1", "text1", date, null)
+        every { dao.getKweet(2) } returns Kweet(2, "user2", "text2", date, null)
+        every { dao.top() } returns listOf(1)
+        every { dao.latest() } returns listOf(2)
+
+        handleRequest(HttpMethod.Get, "/").apply {
+            assertEquals(200, response.status()?.value)
+            assertFalse(response.content!!.contains("There are no kweets yet"))
+            assertTrue(response.content!!.contains("user1"))
+            assertTrue(response.content!!.contains("user2"))
+        }
+
+        verify(exactly = 2) { dao.getKweet(any()) }
+        verify(exactly = 1) { dao.top() }
+        verify(exactly = 1) { dao.latest() }
+    }
+
+    /**
+     * Verifies the behaviour of a login failure. That it should be a redirection to the /user page.
+     */
+    @Test
+    fun testLoginFail() = testApp {
+        handleRequest(HttpMethod.Post, "/login") {
+            addHeader(HttpHeaders.ContentType, ContentType.Application.FormUrlEncoded.toString())
+            setBody(listOf("userId" to "myuser", "password" to "invalid").formUrlEncode())
+        }.apply {
+            assertEquals(302, response.status()?.value)
+        }
+    }
+
+    /**
+     * Verifies a chain of requests verifying the [Login].
+     * It mocks a get [DAOFacade.user] request, checks that posting valid credentials to the /login form
+     * redirects to the user [UserPage] for that user, and reuses the returned cookie for a request
+     * to the [UserPage] and verifies that with that cookie/session, there is a "sign out" text meaning that
+     * the user is logged in.
+     */
+    @Test
+    fun testLoginSuccess() = testApp {
+        val password = "mylongpassword"
+        val passwordHash = hash(password)
+        val sessionCookieName = "SESSION"
+        lateinit var sessionCookie: Cookie
+        every { dao.user("test1", passwordHash) } returns User("test1", "test1@test.com", "test1", passwordHash)
+
+        handleRequest(HttpMethod.Post, "/login") {
+            addHeader(HttpHeaders.ContentType, ContentType.Application.FormUrlEncoded.toString())
+            setBody(listOf("userId" to "test1", "password" to password).formUrlEncode())
+        }.apply {
+            assertEquals(302, response.status()?.value)
+            assertEquals("/user/test1", response.headers["Location"])
+            assertEquals(null, response.content)
+            sessionCookie = response.cookies[sessionCookieName]!!
+        }
+
+        handleRequest(HttpMethod.Get, "/") {
+            addHeader(HttpHeaders.Cookie, "$sessionCookieName=${sessionCookie.value.encodeURLParameter()}")
+        }.apply {
+            assertTrue { response.content!!.contains("sign out") }
+        }
+    }
+
+    /**
+     * A private method used to reduce boilerplate when testing the application.
+     */
+    private fun testApp(callback: TestApplicationEngine.() -> Unit) {
```

Comment:
why not migrate this test class as well?

Reply:
`KweetApplicationTestLegacy` is a class with an old test API. I renamed it from `KweetApplicationTest` and decided to keep it in this sample side-by-side with new tests so that a user can compare old and new APIs in case of migration.